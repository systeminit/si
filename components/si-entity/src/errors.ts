import { findProp } from "si-registry";

export class EntityError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "EntityError";
  }
}

export class InvalidOpPathError extends EntityError {
  constructor(args: { entityType: string; path: string[] }) {
    const message = `Entity type ${args.entityType} does not have a schema entry for path ${args.path}!`;
    super(message);
    this.name = "InvalidOpPathError";
  }
}

export class InvalidContainerPropError extends EntityError {
  constructor(args: { prop: ReturnType<typeof findProp> }) {
    const message = `Prop of type ${args.prop.type} is not a container type, it must be one of: object, map, or array`;
    super(message);
    this.name = "InvalidContainerProp";
  }
}
