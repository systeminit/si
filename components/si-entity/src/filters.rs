use serde::{Deserialize, Serialize};
use tracing_subscriber::fmt::format::FmtSpan;
use warp::Filter;

use si_data::{DataQuery, Db};

use crate::handlers;
use crate::models;

// Backend API
//
// Create - POST /entity/
// Update - PUT /entity/{id}
// Delete - DELETE /entity/{id}
// Get    - GET /entity/{id}
// List   - GET /entity?{query}
//

// This is where you inject the context!
pub fn entities(
    db: &Db,
) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
    entities_list(db.clone())
        .or(entities_get(db.clone()))
        .or(entities_update_field(db.clone()))
}

pub fn entities_list(
    db: Db,
) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
    warp::path!("entities")
        .and(warp::get())
        .and(with_db(db))
        .and(warp::header::<String>("userId"))
        .and(warp::header::<String>("billingAccountId"))
        .and(warp::query::<models::ListQuery>())
        .and_then(handlers::list_entities)
}

pub fn entities_get(
    db: Db,
) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
    warp::path!("entities" / String)
        .and(warp::get())
        .and(with_db(db))
        .and(warp::header::<String>("userId"))
        .and(warp::header::<String>("billingAccountId"))
        .and_then(handlers::get_entity)
}

pub fn entities_update_field(
    db: Db,
) -> impl Filter<Extract = impl warp::Reply, Error = warp::Rejection> + Clone {
    warp::path!("entities" / String)
        .and(warp::patch())
        .and(with_db(db))
        .and(warp::header::<String>("userId"))
        .and(warp::header::<String>("billingAccountId"))
        .and(warp::query::<models::SetField>())
        .and_then(handlers::update_entity_field)
}

fn with_db(db: Db) -> impl Filter<Extract = (Db,), Error = std::convert::Infallible> + Clone {
    warp::any().map(move || db.clone())
}
