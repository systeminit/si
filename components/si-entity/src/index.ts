export {
  ISiEntity,
  SiEntity,
  OpSet,
  OpTombstone,
  OpUnset,
  OpType,
  OpSource,
  EditField,
  CodeDecorationItem,
} from "./siEntity";
export { SiStorable } from "./siStorable";
export {
  Resource,
  SubResource,
  ResourceInternalHealth,
  ResourceInternalStatus,
} from "./resource";
