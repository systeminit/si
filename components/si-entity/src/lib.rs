mod filters;
mod handlers;
mod models;

use si_data::Db;
use si_settings::Settings;
use warp::Filter;

pub async fn start(db: Db, settings: Settings) {
    let api = filters::entities(&db);
    let routes = api
        .with(warp::trace::request())
        .recover(handlers::handle_rejection);
    println!(
        "*** Listening on http://0.0.0.0:{} ***",
        settings.service.port
    );
    warp::serve(routes)
        .run(([0, 0, 0, 0], settings.service.port))
        .await;
}
