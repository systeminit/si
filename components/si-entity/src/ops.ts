import { Inference } from "si-inference";
import { findProp, registry } from "si-registry";
import { v4 as uuidv4 } from "uuid";
import { InvalidOpPathError } from "./errors";
import _ from "lodash";

export enum OpSource {
  Manual = "manual",
  Inferred = "inferred",
}

export interface IOpBase {
  id: string;
  type: string;
  source: OpSource;
  path: string[];
  system: "baseline" | string;
}

export interface OpSetBase extends IOpBase {
  type: "set";
  value:
    | string
    | number
    | boolean
    | Record<string, unknown>
    | Array<unknown>
    | null;
  editPartial?: string;
}

export interface OpSetManual extends OpSetBase {
  source: OpSource.Manual;
  provenance?: never;
}

export interface OpSetInferred extends OpSetBase {
  source: OpSource.Inferred;
  provenance: {
    context: {
      id: string;
      entityType: string;
    }[];
    inference: Inference;
  };
}

export type OpSet = OpSetManual | OpSetInferred;

export function validatePathForEntityType(args: {
  path: OpSet["path"];
  entityType: string;
}): boolean {
  const fullPath = [args.entityType, ...args.path];
  const prop = findProp(fullPath);
  if (prop) {
    return true;
  } else {
    return false;
  }
}

export interface OpSetCreateManualArgs {
  entityType: string;
  system: OpSet["system"];
  path: OpSet["path"];
  value: OpSet["value"];
  editPartial?: string;
}

function createManual({
  entityType,
  value,
  path,
  system,
  editPartial,
}: OpSetCreateManualArgs): OpSetManual {
  const id = uuidv4();
  if (!validatePathForEntityType({ path, entityType })) {
    throw new InvalidOpPathError({ entityType, path });
  }
  return {
    id,
    type: "set",
    source: OpSource.Manual,
    value,
    system,
    path,
    editPartial,
  };
}

export interface OpSetCreateInferredArgs {
  entityType: string;
  system: OpSet["system"];
  path: OpSet["path"];
  value: OpSet["value"];
  editPartial?: OpSet["editPartial"];
  provenance: OpSet["provenance"];
}

function createInferred({
  entityType,
  value,
  path,
  system,
  editPartial,
  provenance,
}: OpSetCreateInferredArgs): OpSetInferred {
  const id = uuidv4();
  if (!validatePathForEntityType({ path, entityType })) {
    throw new InvalidOpPathError({ entityType, path });
  }
  return {
    id,
    type: "set",
    source: OpSource.Inferred,
    value,
    system,
    path,
    editPartial,
    provenance,
  };
}

export interface IsEqualOpSet {
  path: OpSet["path"];
  value: OpSet["value"];
  system: OpSet["system"];
  source: OpSet["source"];
}

function isEqual(left: IsEqualOpSet, right: IsEqualOpSet): boolean {
  const pathIsEqual = _.isEqual(left.path, right.path);
  const valueIsEqual = _.isEqual(left.value, right.value);
  const systemIsEqual = _.isEqual(left.system, right.system);
  const sourceIsEqual = _.isEqual(left.source, right.source);

  return pathIsEqual && valueIsEqual && systemIsEqual && sourceIsEqual;
}

function isEqualExceptValue(left: IsEqualOpSet, right: IsEqualOpSet): boolean {
  const pathIsEqual = _.isEqual(left.path, right.path);
  const systemIsEqual = _.isEqual(left.system, right.system);
  const sourceIsEqual = _.isEqual(left.source, right.source);

  return pathIsEqual && systemIsEqual && sourceIsEqual;
}

export const OpSet = {
  createManual,
  createInferred,
  isEqual,
  isEqualExceptValue,
};
