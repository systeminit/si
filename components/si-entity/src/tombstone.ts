import { OpSource } from "./ops";

export interface Tombstone {
  path: string[];
  source: OpSource;
  system: "baseline" | string;
}
