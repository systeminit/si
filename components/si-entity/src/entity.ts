import { OpSet, OpSource } from "./ops";
import { SiStorable } from "./siStorable";
import { Tombstone } from "./tombstone";
import _ from "lodash";
import { findProp, ItemProp, Prop } from "si-registry";
import { InvalidContainerPropError, InvalidOpPathError } from "./errors";

export interface IEntity {
  id: string;
  entityType: string;
  nodeId: string;
  name: string;
  siStorable: SiStorable;
  dependencies: Record<string, string[]>;
  ops: Array<OpSet>;
  tombstones: Tombstone[];
  properties: {
    [systemId: string]: Record<string, unknown>;
  };
  code: {
    [systemId: string]: string;
  };
}

export interface FindOpForPathAndSystemArgs {
  path: OpSet["path"];
  system: OpSet["system"];
}

function findOpSetForPathAndSystem(
  entity: IEntity,
  { path, system }: FindOpForPathAndSystemArgs,
): OpSet | undefined {
  return _.find(
    entity.ops,
    (o) =>
      _.isEqual(o.path, path) && (o.system == system || o.system == "baseline"),
  );
}

export function emptyValueForContainerProp(
  prop: ReturnType<typeof findProp>,
): Record<string, unknown> | [] {
  if (prop.type == "object") {
    return {};
  } else if (prop.type == "array") {
    return [];
  } else if (prop.type == "map") {
    return {};
  } else {
    throw new InvalidContainerPropError({ prop });
  }
}

export function updateOpSet(entity: IEntity, op: OpSet): void {
  const existingIndex = _.findIndex(entity.ops, (o) =>
    OpSet.isEqualExceptValue(o, op),
  );
  // -1 means there is no op that matches
  if (existingIndex == -1) {
    entity.ops.push(op);
  } else {
    removeDependencies(entity, entity.ops[existingIndex].id);
    entity.ops[existingIndex] = op;
  }
}

export function removeDependencies(entity: IEntity, opSetId: string): IEntity {
  delete entity.dependencies[opSetId];
  return entity;
}

export interface UpdateDependenciesArgs {
  opSetId: string;
  dependencies: string[];
}

export function updateDependencies(
  entity: IEntity,
  { opSetId, dependencies }: UpdateDependenciesArgs,
): IEntity {
  entity.dependencies[opSetId] = dependencies;
  return entity;
}

function listSystemsForEntity(entity: IEntity): string[] {
  const systems: Set<string> = new Set();
  systems.add("baseline");
  for (const opSet of entity.ops) {
    systems.add(opSet.system);
  }
  return Array.from(systems);
}

export interface IsOpSetOverridenArgs {
  opSet: {
    path: OpSet["path"];
    source: OpSet["source"];
    system: OpSet["system"];
    value: OpSet["value"];
  };
  system: string;
}

export function isOpSetOverridden(
  entity: IEntity,
  { opSet, system }: IsOpSetOverridenArgs,
): boolean {
  if (opSet.source == OpSource.Inferred && opSet.system == "baseline") {
    const override = _.find(entity.ops, (toCheckOp) => {
      // The item we are checking is never overriden.
      if (OpSet.isEqual(toCheckOp, opSet)) {
        return false;
      }

      return (
        _.isEqual(toCheckOp.path, opSet.path) &&
        ((toCheckOp.source == OpSource.Manual &&
          toCheckOp.system == "baseline") ||
          toCheckOp.system == system) &&
        !isTombstoned(entity, toCheckOp)
      );
    });
    if (override) {
      return true;
    }
  } else if (opSet.source == OpSource.Inferred && opSet.system != "baseline") {
    const override = _.find(entity.ops, (p) => {
      // Our own object is not an override!
      if (OpSet.isEqual(p, opSet)) {
        return false;
      }

      return (
        _.isEqual(p.path, opSet.path) &&
        p.system == opSet.system &&
        opSet.system == system &&
        !isTombstoned(entity, p)
      );
    });
    if (override) {
      return true;
    }
  } else if (opSet.source == OpSource.Manual && opSet.system == "baseline") {
    const override = _.find(entity.ops, (p) => {
      if (OpSet.isEqual(p, opSet)) {
        return false;
      }
      return (
        _.isEqual(p.path, opSet.path) &&
        p.source == OpSource.Manual &&
        p.system == system &&
        !isTombstoned(entity, p)
      );
    });
    if (override) {
      return true;
    }
  }
  return false;
}

function subPath(long: string[], short: string[]): boolean {
  let checkPath;
  if (long.length == short.length) {
    checkPath = long;
  } else if (long.length > short.length) {
    checkPath = long.slice(0, short.length);
  } else {
    return false;
  }
  return _.isEqual(checkPath, short);
}

export interface IsOpSetTombstonedArgs {
  path: OpSet["path"];
  source: OpSet["source"];
  system: OpSet["system"];
}

export function removeTombstone(
  entity: IEntity,
  tombstone: Tombstone,
): IEntity {
  _.remove(entity.tombstones, (t) => _.isEqual(t, tombstone));
  return entity;
}

export function addTombstone(entity: IEntity, tombstone: Tombstone): IEntity {
  const existingTombstone = _.find(entity.tombstones, (t) =>
    _.isEqual(t, tombstone),
  );
  if (!existingTombstone) {
    entity.tombstones.push(tombstone);
  }
  return entity;
}

export function isTombstoned(
  entity: IEntity,
  { path, source, system }: IsOpSetTombstonedArgs,
): boolean {
  for (const tombstone of entity.tombstones) {
    if (
      subPath(path, tombstone.path) &&
      source == tombstone.source &&
      system == tombstone.system
    ) {
      return true;
    }
  }
  return false;
}

export function generateProperties(entity: IEntity): IEntity {
  //const newProperties: Record<string, unknown> = {};
  //const systems = listSystemsForEntity(entity);
  //for (const system of systems) {
  //  entity.properties[system] = {};
  //}
  //for (const opSet of entity.ops) {
  //  if (opSet.system == "baseline") {
  //    for (const system of systems) {
  //      if (!this.isOverridden(op, system) && !this.isTombstoned(op)) {
  //        _.set(newProperties[system], op.path, _.cloneDeep(op.value));
  //      }
  //    }
  //  }
  //}
  return entity;
}

export interface EntitySetValueArgs {
  path: OpSet["path"];
  system: OpSet["system"];
  value: OpSet["value"];
  editPartial?: OpSet["editPartial"];
  provenance?: OpSet["provenance"];
}

function setValue(
  entity: IEntity,
  { path, system, value, editPartial, provenance }: EntitySetValueArgs,
): IEntity {
  // Walk the path, creating interstitial object as needed, essentially
  // duplicating the auto-viv of lodash

  const dependencies = [];

  if (path.length >= 2) {
    for (let x = 0; x < path.length - 1; x++) {
      const currentPath = path.slice(0, x + 1);
      const existingOp = findOpSetForPathAndSystem(entity, {
        path: currentPath,
        system,
      });
      if (existingOp) {
        dependencies.push(existingOp.id);
      } else {
        const prop = findProp([entity.entityType, ...currentPath]);
        if (!prop) {
          throw new InvalidOpPathError({
            entityType: entity.entityType,
            path: currentPath,
          });
        }
        const emptyValue = emptyValueForContainerProp(prop);
        let dependentOp;
        if (provenance) {
          dependentOp = OpSet.createInferred({
            path: currentPath,
            system,
            value: emptyValue,
            entityType: entity.entityType,
            provenance,
          });
        } else {
          dependentOp = OpSet.createManual({
            path: currentPath,
            system,
            value: emptyValue,
            entityType: entity.entityType,
          });
        }
        updateOpSet(entity, dependentOp);
        updateDependencies(entity, { opSetId: dependentOp.id, dependencies });
      }
    }
  }

  let opSet;
  if (provenance) {
    opSet = OpSet.createInferred({
      path,
      system,
      value,
      editPartial,
      entityType: entity.entityType,
      provenance,
    });
  } else {
    opSet = OpSet.createManual({
      path,
      system,
      value,
      editPartial,
      entityType: entity.entityType,
    });
  }
  updateOpSet(entity, opSet);
  updateDependencies(entity, { opSetId: opSet.id, dependencies });

  return entity;
}

export interface EntitySetManualArgs {
  path: OpSet["path"];
  system: OpSet["system"];
  value: OpSet["value"];
  editPartial?: OpSet["editPartial"];
}

export interface EntitySetInferredArgs {
  path: OpSet["path"];
  system: OpSet["system"];
  value: OpSet["value"];
  editPartial?: OpSet["editPartial"];
  provenance: OpSet["provenance"];
}

function setManualValue(entity: IEntity, args: EntitySetManualArgs): IEntity {
  return setValue(entity, args);
}

function setInferredValue(
  entity: IEntity,
  args: EntitySetInferredArgs,
): IEntity {
  return setValue(entity, args);
}

export const Entity = {
  setManualValue,
  setInferredValue,
  addTombstone,
  removeTombstone,
  findOpSetForPathAndSystem,
};
