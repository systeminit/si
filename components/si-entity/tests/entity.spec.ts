import { findProp } from "si-registry";
import {
  IEntity,
  Entity,
  emptyValueForContainerProp,
  updateOpSet,
  updateDependencies,
  removeDependencies,
  generateProperties,
  isTombstoned,
  isOpSetOverridden,
} from "../src/entity";
import { OpSource, OpSet } from "../src/ops";
import _ from "lodash";
import { Tombstone } from "../src/tombstone";
import { Inference } from "si-inference";

interface EntityCreateArgs {
  entityType: string;
}

const mockInference: Inference = {
  kind: "lambda",
  name: "poop",
  from: [{ targetEntity: true, data: [{ name: true }] }],
  to: { name: true },
  code: "firstEntity.name",
};

const mockContext = [{ id: "1", entityType: "poop" }];

const mockProvenance = {
  inference: mockInference,
  context: mockContext,
};

function createFake({ entityType }: EntityCreateArgs): IEntity {
  return {
    id: "fake",
    entityType,
    nodeId: "fake",
    name: "fake",
    ops: [],
    tombstones: [],
    siStorable: {
      typeName: entityType,
      objectId: "fake",
      billingAccountId: "fake",
      organizationId: "fake",
      workspaceId: "fake",
      tenantIds: ["fake"],
      deleted: false,
    },
    dependencies: {},
    properties: {},
    code: {},
  };
}

describe("Entity", () => {
  describe("updateOpSet", () => {
    test("adds opSet to ops array if it does not exist", () => {
      const entity = createFake({ entityType: "torture" });
      const opSet = OpSet.createManual({
        entityType: "torture",
        value: "death",
        path: ["standardString"],
        system: "baseline",
      });
      updateOpSet(entity, opSet);
      expect(entity.ops).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            value: "death",
          }),
        ]),
      );
    });

    test("updates the existing opSet if the path and system match", () => {
      const entity = createFake({ entityType: "torture" });

      const opSet = OpSet.createManual({
        entityType: "torture",
        value: "death",
        path: ["standardString"],
        system: "baseline",
      });
      updateOpSet(entity, opSet);

      const mastodonOpSet = OpSet.createManual({
        entityType: "torture",
        value: "mastodon",
        path: ["standardString"],
        system: "baseline",
      });
      updateOpSet(entity, mastodonOpSet);

      expect(entity.ops).not.toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            value: "death",
          }),
        ]),
      );

      expect(entity.ops).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            value: "mastodon",
          }),
        ]),
      );
    });
  });

  describe("removeDependencies", () => {
    test("removes dependency array for given id", () => {
      const entity = createFake({ entityType: "torture" });
      updateDependencies(entity, { opSetId: "1", dependencies: ["2", "3"] });
      expect(entity.dependencies).toEqual(
        expect.objectContaining({
          "1": ["2", "3"],
        }),
      );
      removeDependencies(entity, "1");
      expect(entity.dependencies).not.toEqual(
        expect.objectContaining({
          "1": ["2", "3"],
        }),
      );
    });
  });

  describe("updateDependencies", () => {
    test("sets the dependencies", () => {
      const entity = createFake({ entityType: "torture" });
      updateDependencies(entity, { opSetId: "1", dependencies: ["2", "3"] });
      expect(entity.dependencies).toEqual(
        expect.objectContaining({
          "1": ["2", "3"],
        }),
      );
    });
  });

  describe("emptyValueForContainerProp", () => {
    test("returns {} for object", () => {
      const prop = findProp(["torture", "nestedTest"]);
      const value = emptyValueForContainerProp(prop);
      expect(value).toEqual({});
    });
    test("returns [] for arrays", () => {
      const prop = findProp(["torture", "complexArray"]);
      const value = emptyValueForContainerProp(prop);
      expect(value).toEqual([]);
    });
    test("returns {} for maps", () => {
      const prop = findProp(["torture", "mappy"]);

      const value = emptyValueForContainerProp(prop);
      expect(value).toEqual({});
    });
    test("throws an error on non container prop types", () => {
      const prop = findProp(["torture", "standardString"]);
      expect(() => {
        emptyValueForContainerProp(prop);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe("findOpSetForPathAndSystem", () => {
    test("returns undefined if not found", () => {
      const entity = createFake({ entityType: "torture" });
      const result = Entity.findOpSetForPathAndSystem(entity, {
        path: ["shoes"],
        system: "baseline",
      });
      expect(result).toBe(undefined);
    });
    test("returns opSet if found", () => {
      const entity = createFake({ entityType: "torture" });
      Entity.setManualValue(entity, {
        path: ["standardString"],
        system: "baseline",
        value: "death",
      });
      const result = Entity.findOpSetForPathAndSystem(entity, {
        path: ["standardString"],
        system: "baseline",
      });
      expect(result).toEqual(expect.objectContaining({ value: "death" }));
    });
    test("returns a baseline opSet if found when asking for a specific system", () => {
      const entity = createFake({ entityType: "torture" });
      Entity.setManualValue(entity, {
        path: ["standardString"],
        system: "baseline",
        value: "death",
      });
      const result = Entity.findOpSetForPathAndSystem(entity, {
        path: ["standardString"],
        system: "verge",
      });
      expect(result).toEqual(expect.objectContaining({ value: "death" }));
    });
  });

  describe("setManualValue", () => {
    test("adds an op for the manual value", () => {
      const entity = createFake({ entityType: "torture" });
      const result = Entity.setManualValue(entity, {
        path: ["standardString"],
        system: "baseline",
        value: "death",
      });
      expect(result).toBe(entity);
      expect(result.ops).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            path: ["standardString"],
            system: "baseline",
            value: "death",
          }),
        ]),
      );
      const op = _.find(result.ops, (o) => o.value == "death");
      const depObj: Record<string, string[]> = {};
      depObj[op.id] = [];
      expect(result.dependencies).toEqual(expect.objectContaining(depObj));
    });
    describe("nested values", () => {
      describe("object with nested string", () => {
        test("adds an op for each missing nested value along with the final value", () => {
          const entity = createFake({ entityType: "torture" });
          const result = Entity.setManualValue(entity, {
            path: ["nestedTest", "squirrel", "name"],
            system: "baseline",
            value: "monkey",
          });
          expect(result.ops).toEqual(
            expect.arrayContaining([
              expect.objectContaining({
                path: ["nestedTest"],
                system: "baseline",
                value: {},
              }),
              expect.objectContaining({
                path: ["nestedTest", "squirrel"],
                system: "baseline",
                value: {},
              }),
              expect.objectContaining({
                path: ["nestedTest", "squirrel", "name"],
                system: "baseline",
                value: "monkey",
              }),
            ]),
          );
        });
      });
    });
  });

  describe("addTombstone", () => {
    test("adds tombstone when no matching tombstoned exists", () => {
      const entity = createFake({ entityType: "torture" });
      const tombstone: Tombstone = {
        path: ["standardString"],
        system: "baseline",
        source: OpSource.Manual,
      };
      Entity.addTombstone(entity, tombstone);
      expect(entity.tombstones).toEqual(
        expect.arrayContaining([expect.objectContaining(tombstone)]),
      );
    });

    test("does not add duplicate tombstones", () => {
      const entity = createFake({ entityType: "torture" });
      const tombstone: Tombstone = {
        path: ["standardString"],
        system: "baseline",
        source: OpSource.Manual,
      };
      Entity.addTombstone(entity, tombstone);
      Entity.addTombstone(entity, tombstone);
      Entity.addTombstone(entity, tombstone);
      expect(entity.tombstones).toEqual(
        expect.arrayContaining([expect.objectContaining(tombstone)]),
      );
      expect(entity.tombstones.length).toBe(1);
    });
  });

  describe("removeTombstone", () => {
    test("removes tombstone from the list", () => {
      const entity = createFake({ entityType: "torture" });
      const tombstone: Tombstone = {
        path: ["standardString"],
        system: "baseline",
        source: OpSource.Manual,
      };
      Entity.addTombstone(entity, tombstone);
      expect(entity.tombstones).toEqual(
        expect.arrayContaining([expect.objectContaining(tombstone)]),
      );
      Entity.removeTombstone(entity, tombstone);
      expect(entity.tombstones).not.toEqual(
        expect.arrayContaining([expect.objectContaining(tombstone)]),
      );
    });

    test("does nothing when tombstone is already gone", () => {
      const entity = createFake({ entityType: "torture" });
      const tombstone: Tombstone = {
        path: ["standardString"],
        system: "baseline",
        source: OpSource.Manual,
      };
      Entity.addTombstone(entity, tombstone);
      expect(entity.tombstones).toEqual(
        expect.arrayContaining([expect.objectContaining(tombstone)]),
      );
      Entity.removeTombstone(entity, tombstone);
      expect(entity.tombstones).not.toEqual(
        expect.arrayContaining([expect.objectContaining(tombstone)]),
      );
      Entity.removeTombstone(entity, tombstone);
      expect(entity.tombstones).not.toEqual(
        expect.arrayContaining([expect.objectContaining(tombstone)]),
      );
    });
  });

  describe("isTombstoned", () => {
    test("returns false it is not tombstoned", () => {
      const entity = createFake({ entityType: "torture" });
      const op = {
        path: ["standardString"],
        system: "baseline",
        value: "metallica",
      };
      Entity.setManualValue(entity, op);
      const result = isTombstoned(entity, { source: OpSource.Manual, ...op });
      expect(result).toBe(false);
    });

    test("returns false with tombstones that do not match", () => {
      const entity = createFake({ entityType: "torture" });
      const op = {
        path: ["standardString"],
        system: "baseline",
        value: "metallica",
      };
      const stringArrayTombstone = {
        ...op,
        source: OpSource.Manual,
        path: ["stringArray"],
      };
      Entity.addTombstone(entity, stringArrayTombstone);
      const inferredTombstone = { ...op, source: OpSource.Inferred };
      Entity.addTombstone(entity, inferredTombstone);
      Entity.setManualValue(entity, op);
      const result = isTombstoned(entity, { ...op, source: OpSource.Manual });
      expect(result).toBe(false);
    });

    test("returns true if tombstoned by exact path", () => {
      const entity = createFake({ entityType: "torture" });
      const op = {
        path: ["standardString"],
        system: "baseline",
        value: "metallica",
      };
      Entity.setManualValue(entity, op);
      const tombstone = { source: OpSource.Manual, ...op };
      Entity.addTombstone(entity, tombstone);
      const result = isTombstoned(entity, tombstone);
      expect(result).toBe(true);
    });

    test("returns true if tombstoned by sub path", () => {
      const entity = createFake({ entityType: "torture" });
      const op = {
        path: ["nestedTest", "squirrel", "name"],
        system: "baseline",
        value: "metallica",
      };
      Entity.setManualValue(entity, op);
      const tombstone = {
        source: OpSource.Manual,
        path: ["nestedTest"],
        ...op,
      };
      Entity.addTombstone(entity, tombstone);
      const result = isTombstoned(entity, tombstone);
      expect(result).toBe(true);
    });
  });

  describe("isOpSetOverriden", () => {
    describe("inferred", () => {
      describe("baseline", () => {
        test("is not overriden when no other values exist", () => {
          const entity = createFake({ entityType: "torture" });
          const op = {
            path: ["standardString"],
            system: "baseline",
            value: "metallica",
            provenance: mockProvenance,
          };
          Entity.setInferredValue(entity, op);
          const result = isOpSetOverridden(entity, {
            opSet: { ...op, source: OpSource.Inferred },
            system: "baseline",
          });
          expect(result).toBe(false);
        });
        test("is overriden when an inferred system value exists", () => {
          const entity = createFake({ entityType: "torture" });
          const ops = [
            {
              path: ["standardString"],
              system: "baseline",
              value: "metallica",
              provenance: mockProvenance,
            },
            {
              path: ["standardString"],
              system: "system:1",
              value: "cuba",
              provenance: mockProvenance,
            },
          ];
          for (const op of ops) {
            Entity.setInferredValue(entity, op);
          }

          const baselineResult = isOpSetOverridden(entity, {
            opSet: { ...ops[0], source: OpSource.Inferred },
            system: "baseline",
          });
          expect(baselineResult).toBe(false);

          const systemResult = isOpSetOverridden(entity, {
            opSet: { ...ops[0], source: OpSource.Inferred },
            system: "system:1",
          });
          expect(systemResult).toBe(true);
        });
        test("is overriden when a manual baseline value exists", () => {
          const entity = createFake({ entityType: "torture" });
          const ops = [
            {
              path: ["standardString"],
              system: "baseline",
              value: "metallica",
              provenance: mockProvenance,
            },
          ];
          for (const op of ops) {
            Entity.setInferredValue(entity, op);
          }
          const manualOp = {
            path: ["standardString"],
            system: "baseline",
            value: "cuba",
          };
          Entity.setManualValue(entity, manualOp);

          const baselineResult = isOpSetOverridden(entity, {
            opSet: { ...ops[0], source: OpSource.Inferred },
            system: "baseline",
          });
          expect(baselineResult).toBe(true);
        });
      });

      describe("system", () => {
        test("is not overriden when no other values exist", () => {
          const entity = createFake({ entityType: "torture" });
          const op = {
            path: ["standardString"],
            system: "system:1",
            value: "metallica",
            provenance: mockProvenance,
          };
          Entity.setInferredValue(entity, op);
          const result = isOpSetOverridden(entity, {
            opSet: { ...op, source: OpSource.Inferred },
            system: "system:1",
          });
          expect(result).toBe(false);
        });
        test("is not overriden when an inferred baseline value exists", () => {
          const entity = createFake({ entityType: "torture" });
          const baselineOp = {
            path: ["standardString"],
            system: "baseline",
            value: "alice",
            provenance: mockProvenance,
          };
          Entity.setInferredValue(entity, baselineOp);
          const op = {
            path: ["standardString"],
            system: "system:1",
            value: "metallica",
            provenance: mockProvenance,
          };
          Entity.setInferredValue(entity, op);
          const result = isOpSetOverridden(entity, {
            opSet: { ...op, source: OpSource.Inferred },
            system: "system:1",
          });
          expect(result).toBe(false);
        });
        test("is overriden when a manual value exists for the same system", () => {
          const entity = createFake({ entityType: "torture" });
          const manualOp = {
            path: ["standardString"],
            system: "system:1",
            value: "alice",
          };
          Entity.setManualValue(entity, manualOp);
          const op = {
            path: ["standardString"],
            system: "system:1",
            value: "metallica",
            provenance: mockProvenance,
          };
          Entity.setInferredValue(entity, op);

          const result = isOpSetOverridden(entity, {
            opSet: { ...op, source: OpSource.Inferred },
            system: "system:1",
          });
          expect(result).toBe(true);
        });
      });
    });
    describe.only("manual", () => {
      describe("baseline", () => {
        test("is not overriden when no other values exist", () => {
          const entity = createFake({ entityType: "torture" });
          const op = {
            path: ["standardString"],
            system: "baseline",
            value: "metallica",
          };
          Entity.setManualValue(entity, op);
          const result = isOpSetOverridden(entity, {
            opSet: { ...op, source: OpSource.Manual },
            system: "baseline",
          });
          expect(result).toBe(false);
        });
        test("is not overriden when inferred values also exist", () => {
          const entity = createFake({ entityType: "torture" });
          const op = {
            path: ["standardString"],
            system: "baseline",
            value: "metallica",
          };
          Entity.setManualValue(entity, op);

          const inferredOp = {
            path: ["standardString"],
            system: "baseline",
            value: "metallica",
            provenance: mockProvenance,
          };
          Entity.setInferredValue(entity, inferredOp);

          const result = isOpSetOverridden(entity, {
            opSet: { ...op, source: OpSource.Manual },
            system: "baseline",
          });
          expect(result).toBe(false);
        });
        test("is overriden when a manual system value exists", () => {
          const entity = createFake({ entityType: "torture" });
          const op = {
            path: ["standardString"],
            system: "baseline",
            value: "metallica",
          };
          Entity.setManualValue(entity, op);

          const manualOp = {
            path: ["standardString"],
            system: "system:1",
            value: "metallica",
          };
          Entity.setManualValue(entity, manualOp);

          const result = isOpSetOverridden(entity, {
            opSet: { ...op, source: OpSource.Manual },
            system: "system:1",
          });
          expect(result).toBe(true);
        });
        test("is overridden when an inferred system value exists", () => {
          const entity = createFake({ entityType: "torture" });
          const op = {
            path: ["standardString"],
            system: "baseline",
            value: "metallica",
          };
          Entity.setManualValue(entity, op);

          const inferredOp = {
            path: ["standardString"],
            system: "system:1",
            value: "metallica",
            provenance: mockProvenance,
          };
          Entity.setInferredValue(entity, inferredOp);

          const result = isOpSetOverridden(entity, {
            opSet: { ...op, source: OpSource.Manual },
            system: "system:1",
          });
          expect(result).toBe(true);
        });
      });
    });
  });

  //describe("generateProperties", () => {
  //  describe("no tombstones", () => {
  //    describe("baseline values", () => {
  //      test("sets a string value for a root path", () => {
  //        const entity = createFake({ entityType: "torture" });
  //        Entity.setManualValue(entity, {
  //          path: ["standardString"],
  //          system: "baseline",
  //          value: "metallica",
  //        });
  //        const result = generateProperties(entity);
  //        expect(result.properties).toEqual(
  //          expect.objectContaining({
  //            baseline: {
  //              standardString: "metallica",
  //            },
  //          }),
  //        );
  //      });
  //    });
  //  });
  //});
});
