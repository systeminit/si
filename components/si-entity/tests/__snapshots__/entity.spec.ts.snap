// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity emptyValueForContainerProp throws an error on non container prop types 1`] = `"Prop of type string is not a container type, it must be one of: object, map, or array"`;
